File: src/main/java/com/akagiyui/drive/model/response/UserInfoResponse.java
package com.akagiyui.drive.model.response;

import com.akagiyui.drive.entity.Role;
import com.akagiyui.drive.entity.User;
import com.akagiyui.drive.model.Permission;
import lombok.Data;
import lombok.experimental.Accessors;

import java.util.Date;
import java.util.List;

/**
 * 用户信息响应
 *
 * @author AkagiYui
 */
@Data
@Accessors(chain = true)
public class UserInfoResponse {
    /**
     * 用户id
     */
    private String id;
    /**
     * 用户名
     */
    private String username;
    /**
     * 昵称
     */
    private String nickname;
    /**
     * 邮箱
     */
    private String email;
    /**
     * 已禁用
     */
    private boolean disabled;
    /**
     * 注册时间
     */
    private Date registerTime;

    /**
     * 权限
     */
    private List<String> permissions;

    public UserInfoResponse(User user) {
        this.id = user.getId();
        this.username = user.getUsername();
        this.nickname = user.getNickname();
        this.email = user.getEmail();
        this.disabled = user.getDisabled();
        this.registerTime = user.getCreateTime();

        List<Role> roles = user.getRoles();
        List<Permission> allPermissions = roles.stream().flatMap(role -> role.getPermissions().stream()).toList();
        this.permissions = allPermissions.stream().map(Permission::name).toList();
    }

    public static List<UserInfoResponse> fromUserList(List<User> users) {
        return users.stream().map(UserInfoResponse::new).toList();
    }
}

File: app/src/main/kotlin/com/akagiyui/drive/model/response/UserInfoResponse.kt
package com.akagiyui.drive.model.response

import com.akagiyui.drive.entity.User

/**
 * 用户信息响应
 *
 * @author AkagiYui
 */
data class UserInfoResponse(
    /**
     * 用户ID
     */
    val id: String,

    /**
     * 用户名
     */
    val username: String,

    /**
     * 昵称
     */
    val nickname: String,

    /**
     * 邮箱
     */
    val email: String,

    /**
     * 已禁用
     */
    val disabled: Boolean,

    /**
     * 注册时间
     */
    val registerTime: Long,

    /**
     * 权限
     */
    val permissions: List<String>,
) {
    constructor(user: User) : this(
        id = user.id,
        username = user.username,
        nickname = user.nickname ?: user.username,
        email = user.email ?: "",
        disabled = user.disabled,
        registerTime = user.createTime.time,
        permissions = user.roles.flatMap { it.permissions }.map { it.name }
    )
}

/**
 * 从用户列表转换
 *
 * @return 用户信息响应列表
 */
fun List<User>.toResponse(): List<UserInfoResponse> {
    return this.map { UserInfoResponse(it) }.toList()
}
